### PR TITLE
A lot of changes, please review and choose

### PR DESCRIPTION
**Pull request by gakowalski**
Hello, I've modified your code recently and made some number of changes that might be useful but also some might not be good, so please, review **carefully** and eventually cherry-pick good commits.

Notable good changes:
* Added suggested PSQL config for a tile server - this **MASSIVELY** improved overall performance

Good changes:
* Install *latest* PostgreSQL 10 available;
* Turn off autovacuum while importing data - few people reported that it can greatly improve import performance;
* Multithread make (to speed up compilation);
* Gracefully stop PostgreSQL at the end of `run.sh` to prevent DB corruption.
* Updated Lua to 5.3

Maybe not so good decisions:
* Apache logs are redirected to stdout and stderr, so you can view them merged with logs from renderd (I've found this useful for debugging)

Controversial changes:
* Dropped a lot of GPG and SSL verifications (probably easy to fix)
* Made a lot of edits which substantially change the Dockerfile to achieve some optimizations but which might make everything less maintainable and harder to debug

### CODE REVIEW

**Comment by Overv** [comment]
This is looking really nice and I'd be happy to merge this after some modifications. The optimization of the Dockerfile is awesome and not really controversial as far as I'm considered. However, I do have to ask what the reasoning behind the SSL related changes like `GIT_SSL_NO_VERIFY=true`, `npm config set strict-ssl=false` and `ENV PYTHONHTTPSVERIFY=0` is? A corporate proxy? :)

**Comment by gakowalski** [comment]
When sorting packages installed by apt-get and removing redundant or unneeded ones, I probably threw one that indirectly installed packages related to SSL. Or maybe packages from PostgreSQL had other dependencies in the official Ubuntu distribution and others in the manufacturer's repository? This was the case with PostGIS. I didn't want to solve that riddle at that time because I already had more interesting ideas for other changes and experiments, so I left it for later, although it's probably a trivial matter. Now that I think about it, maybe the solution to the problem is:
`apt-get install -y apt-transport-https ca-certificates`

I'll check it in a moment.

**Comment by gakowalski** [comment]
I checked it, it works. :-) I have merged it into my master branch, so it should appear here as a part of the pull request.

**Comment by Overv** [comment]
Cool, I will test this PR this week and then merge it.

**Comment by gakowalski** [comment]
Ah, one thing that is *not* in this pull request is a major upgrade of PostgreSQL from 10 to 11. It's very easy to do it and it works well for a "clean install" with empty docker volume but I don't know how to implement database migration with `pg_upgrade` for a case of existing PSQL 10 database so I'm not introducing it into `master` branch. Also, I didn't see much difference in performance, so maybe it's not worth the time. Anyway, if you're interested, check this commit and try for yourself (and remember to update paths in the Makefile):
https://github.com/pot-gov-pl/openstreetmap-tile-server/commit/110a540c99e32ad3efa630a37e1d113af6edc5d0

**Comment by gakowalski** [comment]
One more thing added to this pull request: i've updated README.md on how to deal with this error:

> renderd[126]: ERROR: failed to render TILE ajt 6 32-39 16-23,
> renderd[126]: reason: Postgis Plugin: ERROR: could not resize shared memory segment 

It's described in detail here: https://github.com/pot-gov-pl/openstreetmap-tile-server/issues/3

**Comment by Istador** [comment]
Please enable the autovacuum again when `run`ning the server or disable it only temporary during the `import`.

The way it's currently written, disables the autovacuum during `import` permanently. Because the config file is modified during this `import` step and the modified config file is kept inside the `openstreetmap-data` volume. This affects posgresql when using `run`.

---

[Tuning Your PostgreSQL Server](https://wiki.postgresql.org/wiki/Tuning_Your_PostgreSQL_Server)

> The autovacuum process takes care of several maintenance chores inside your database that you really need. Generally, if you think you need to turn regular vacuuming off because it's taking too much time or resources, that means you're doing it wrong. The answer to almost all vacuuming problems is to vacuum more often, not less, so that each individual vacuum operation has less to clean up.
> 
> However, it's acceptable to disable autovacuum for short periods of time, for instance when bulk loading large amounts of data.

**Comment by Istador** [comment]
Please enable the autovacuum again when `run`ning the server or disable it only temporary during the `import`.

The way it's currently written, disables the autovacuum during `import` permanently. Because the config file is modified during this `import` step and the modified config file is kept inside the container. When the container isn't thrown away but reused (e.g. when using `docker start` or `docker-compose`), this affects posgresql when using `run` on the same container.

I think the best way to handle this would be to put not one but two config files inside this repository.
One that is optimised for the `import` and one that is optimized for `run`ning the server.
The `run.sh` would then choose which one to use before starting postgresql.
It's also easier for the user of the image to overwrite the config files with his own settings, if its content isn't modified by the script (maybe he wants autovacuum=on during import because otherwise disk space might be a problem for him).

---

[Tuning Your PostgreSQL Server](https://wiki.postgresql.org/wiki/Tuning_Your_PostgreSQL_Server)

> The autovacuum process takes care of several maintenance chores inside your database that you really need. Generally, if you think you need to turn regular vacuuming off because it's taking too much time or resources, that means you're doing it wrong. The answer to almost all vacuuming problems is to vacuum more often, not less, so that each individual vacuum operation has less to clean up.
> 
> However, it's acceptable to disable autovacuum for short periods of time, for instance when bulk loading large amounts of data.

**Comment by Overv** [comment]
@Istador That is a good idea.

@gakowalski Could you change the config to have a separate one for import/run and pick one based on the command? That way we can get rid of hacks like writing `autovacuum` to it at runtime and allow users to override them through mounts.

**Comment by stevo01** [comment]
> @gakowalski Could you change the config to have a separate one for import/run and pick one based on the command? That way we can get rid of hacks like writing `autovacuum` to it at runtime and allow users to override them through mounts.

I created Pull Request which allows the user to enable/disable autovacuum with environment variable. Hope this allows you to continue integration of the "gakowalski" patches.

File: postgresql.custom.conf [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+max_wal_size = 2GB
+work_mem = 32MB
+maintenance_work_mem = 256MB
+
+# Suggested settings from
+# https://github.com/openstreetmap/chef/blob/master/roles/tile.rb#L38-L45
+
+max_connections = 250
+temp_buffers = 32MB
+work_mem = 128MB
```
Comment (stevo01) [inline postgresql.custom.conf]: you define "work_mem" twize. (see line 7)

File: postgresql.custom.conf [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+wal_buffers = 1024kB
+wal_writer_delay = 500ms
+commit_delay = 10000
+# checkpoint_segments = 60 # unrecognized in psql 10.7.1
+max_wal_size = 2880MB
```
Comment (stevo01) [inline postgresql.custom.conf]: you define "work_mem" twize. (see line 6)

Comment (Overv) [reply postgresql.custom.conf]: I've fixed this in a separate commit.

**Comment by Overv** [comment]
I've tested your changes along with the contribution by @stevo01 and it works great! I've merged the changes and pushed a new image to Docker Hub. Thank you both very much :+1:

**Pull request merged**